crates/prover/src/adapter/memory.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::HashMap;
+
+use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fields::qm31::QM31;
+
+use crate::adapter::instructions::MemoryArg;
+
+#[derive(Copy, Clone, Default, Debug)]
+pub struct MemoryEntry {
+    pub address: u32,
+    pub value: [u32; 4],
+}
+
+impl From<crate::adapter::io::IoMemoryEntry> for MemoryEntry {
+    fn from(io_entry: crate::adapter::io::IoMemoryEntry) -> Self {
+        Self {
+            address: io_entry.address,
+            value: io_entry.value,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Default, Debug)]
+pub struct TraceEntry {
+    pub pc: u32,
+    pub fp: u32,
+}
+
+impl From<crate::adapter::io::IoTraceEntry> for TraceEntry {
+    fn from(io_entry: crate::adapter::io::IoTraceEntry) -> Self {
+        Self {
+            pc: io_entry.pc,
+            fp: io_entry.fp,
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct MemoryBoundaries {
+    pub initial_memory: Vec<(M31, QM31, M31)>,
+    pub final_memory: Vec<(M31, QM31, M31)>,
+}
+
+#[derive(Debug, Default)]
+pub struct MemoryCache {
+    clock_cache: HashMap<u32, u32>,
```

Comment:
I think that you don't need the clock cache, just insert into the current_memory to get the previous value. if it's none, then append also to the `initial_memory`